Add BSF and BSR fallbacks for BitOperations methods

saucecontrol opened the pull request:
Testing the assertion @tannergooding made in https://github.com/dotnet/runtime/issues/953#issuecomment-608949008 that adding x86 intrinsics is as easy as 1,2,3,4,5,6,7 😄

I wanted to get feedback on the introduction of an X86Base ISA to mirror what we have on ARM, and an `X86Base` class in S.R.I to house the methods for new instructions.  I recall @jkotas mentioning the possibility of exposing some of the x86 base instructions as internal HWIntrinsics that could be used from higher-level methods, although I can't seem to find the discussion now.

I've added HWIntrinsics implementations of `BSF` and `BSR` and used them as fallbacks in some of the `BitOperations` methods that use `LZCNT` and `TZCNT`.  Since BMI1 requires VEX encoding, it won't be supported on older processors or newer non-VEX processors like the Intel Atom line.  The fallbacks will also improve codegen in R2R images, where VEX encoding is disabled.

Benchmark results with BMI1 and LZCNT disabled:

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Xeon CPU E3-1505M v6 3.00GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.300-preview-015048
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.16006, CoreFX 5.0.20.16006), X64 RyuJIT
  Job-XPOYMR : .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT
  Job-ROWHAK : .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:DebugType=portable  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

```
|                  Method |                    Toolchain |       Mean |     Error |    StdDev |     Median |        Min |        Max | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------ |----------------------------- |-----------:|----------:|----------:|-----------:|-----------:|-----------:|------:|--------:|------:|------:|------:|----------:|
|   LeadingZeroCount_uint | \corerun-bsf-bsr\CoreRun.exe | 1,176.0 ns |   8.28 ns |   7.74 ns | 1,176.4 ns | 1,161.6 ns | 1,188.7 ns |  0.30 |    0.01 |     - |     - |     - |         - |
|   LeadingZeroCount_uint |  \corerun-master\CoreRun.exe | 3,867.1 ns | 172.27 ns | 198.39 ns | 3,855.8 ns | 3,509.7 ns | 4,214.0 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |           |           |            |            |            |       |         |       |       |       |           |
|  LeadingZeroCount_ulong | \corerun-bsf-bsr\CoreRun.exe | 1,156.3 ns |  22.38 ns |  23.94 ns | 1,158.5 ns | 1,115.4 ns | 1,204.8 ns |  0.36 |    0.01 |     - |     - |     - |         - |
|  LeadingZeroCount_ulong |  \corerun-master\CoreRun.exe | 3,185.6 ns |  73.96 ns |  82.21 ns | 3,178.5 ns | 3,091.1 ns | 3,362.5 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |           |           |            |            |            |       |         |       |       |       |           |
|               Log2_uint | \corerun-bsf-bsr\CoreRun.exe |   883.6 ns |   9.93 ns |   9.29 ns |   880.6 ns |   871.2 ns |   903.2 ns |  0.31 |    0.01 |     - |     - |     - |         - |
|               Log2_uint |  \corerun-master\CoreRun.exe | 2,806.4 ns |  44.82 ns |  41.93 ns | 2,807.6 ns | 2,730.5 ns | 2,872.6 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |           |           |            |            |            |       |         |       |       |       |           |
|              Log2_ulong | \corerun-bsf-bsr\CoreRun.exe |   875.5 ns |  12.38 ns |  11.58 ns |   876.3 ns |   856.2 ns |   896.4 ns |  0.22 |    0.00 |     - |     - |     - |         - |
|              Log2_ulong |  \corerun-master\CoreRun.exe | 4,058.1 ns |  42.23 ns |  39.51 ns | 4,059.0 ns | 4,001.2 ns | 4,138.1 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |           |           |            |            |            |       |         |       |       |       |           |
|  TrailingZeroCount_uint | \corerun-bsf-bsr\CoreRun.exe |   885.0 ns |   8.74 ns |   7.30 ns |   885.3 ns |   874.4 ns |   897.3 ns |  0.52 |    0.08 |     - |     - |     - |         - |
|  TrailingZeroCount_uint |  \corerun-master\CoreRun.exe | 1,710.8 ns | 206.28 ns | 237.55 ns | 1,607.4 ns | 1,509.7 ns | 2,174.5 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |           |           |            |            |            |       |         |       |       |       |           |
| TrailingZeroCount_ulong | \corerun-bsf-bsr\CoreRun.exe | 1,030.8 ns |  19.25 ns |  19.76 ns | 1,036.3 ns |   984.6 ns | 1,061.7 ns |  0.69 |    0.02 |     - |     - |     - |         - |
| TrailingZeroCount_ulong |  \corerun-master\CoreRun.exe | 1,484.5 ns |  45.58 ns |  52.49 ns | 1,470.6 ns | 1,425.7 ns | 1,625.9 ns |  1.00 |    0.00 |     - |     - |     - |         - |

And enabled:

|                  Method |                    Toolchain |       Mean |    Error |   StdDev |     Median |        Min |        Max | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------ |----------------------------- |-----------:|---------:|---------:|-----------:|-----------:|-----------:|------:|--------:|------:|------:|------:|----------:|
|   LeadingZeroCount_uint | \corerun-bsf-bsr\CoreRun.exe |   632.8 ns |  8.39 ns |  7.85 ns |   630.3 ns |   624.2 ns |   649.7 ns |  0.99 |    0.02 |     - |     - |     - |         - |
|   LeadingZeroCount_uint |  \corerun-master\CoreRun.exe |   636.9 ns |  8.66 ns |  8.10 ns |   636.2 ns |   623.9 ns |   653.7 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |          |          |            |            |            |       |         |       |       |       |           |
|  LeadingZeroCount_ulong | \corerun-bsf-bsr\CoreRun.exe |   634.0 ns |  8.36 ns |  7.82 ns |   632.2 ns |   624.7 ns |   649.9 ns |  1.00 |    0.01 |     - |     - |     - |         - |
|  LeadingZeroCount_ulong |  \corerun-master\CoreRun.exe |   631.2 ns |  8.40 ns |  7.86 ns |   629.3 ns |   616.1 ns |   643.2 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |          |          |            |            |            |       |         |       |       |       |           |
|               Log2_uint | \corerun-bsf-bsr\CoreRun.exe |   729.5 ns | 14.35 ns | 12.72 ns |   728.4 ns |   711.4 ns |   751.3 ns |  0.70 |    0.05 |     - |     - |     - |         - |
|               Log2_uint |  \corerun-master\CoreRun.exe | 1,021.8 ns | 62.09 ns | 71.51 ns | 1,005.0 ns |   926.7 ns | 1,175.7 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |          |          |            |            |            |       |         |       |       |       |           |
|              Log2_ulong | \corerun-bsf-bsr\CoreRun.exe |   735.5 ns |  7.79 ns |  7.29 ns |   737.1 ns |   720.3 ns |   748.9 ns |  0.42 |    0.01 |     - |     - |     - |         - |
|              Log2_ulong |  \corerun-master\CoreRun.exe | 1,743.7 ns | 33.28 ns | 34.18 ns | 1,739.0 ns | 1,684.3 ns | 1,821.5 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |          |          |            |            |            |       |         |       |       |       |           |
|  TrailingZeroCount_uint | \corerun-bsf-bsr\CoreRun.exe |   626.5 ns |  9.07 ns |  8.48 ns |   625.5 ns |   614.8 ns |   640.9 ns |  0.99 |    0.02 |     - |     - |     - |         - |
|  TrailingZeroCount_uint |  \corerun-master\CoreRun.exe |   633.5 ns |  8.48 ns |  7.93 ns |   631.2 ns |   622.1 ns |   648.7 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                         |                              |            |          |          |            |            |            |       |         |       |       |       |           |
| TrailingZeroCount_ulong | \corerun-bsf-bsr\CoreRun.exe |   636.7 ns | 11.54 ns | 10.79 ns |   632.3 ns |   621.0 ns |   661.9 ns |  1.01 |    0.02 |     - |     - |     - |         - |
| TrailingZeroCount_ulong |  \corerun-master\CoreRun.exe |   630.5 ns |  7.80 ns |  7.29 ns |   627.5 ns |   620.2 ns |   648.0 ns |  1.00 |    0.00 |     - |     - |     - |         - |

Note the nice little improvement to `Log2` by replacing a subtraction with xor to reverse the index returned by `LZCNT`

tannergooding commented:
I probably won't get to reviewing this before Monday, but assigning myself so I don't forget 😄

saucecontrol committed:
add X86Base to Utf8String S.R.I. shim

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -67,12 +67,20 @@ public static int LeadingZeroCount(uint value)
+                return 31 ^ (int)X86Base.BitScanReverse(value);
+            }
+
             return 31 - Log2SoftwareFallback(value);
```
Do the xor trick here as well?

saucecontrol replied:
Sure thing.  Also could apply to the ARM versions, but I don't know enough about their codegen to know if it makes a difference there.  Certainly couldn't hurt.

tannergooding commented:
Looks like something isn't quite hooked up right as `X86Base.IsSupported` is returning `false` for various tests.

tannergooding commented:
CC. @CarolEidt and @echesakovMSFT 

Also CC. @davidwrighton to help validate the new `IsSupported` check is hooked up correctly

tannergooding commented on src/coreclr/src/inc/corinfoinstructionset.h:
```diff
@@ -264,6 +276,8 @@ inline CORINFO_InstructionSetFlags EnsureInstructionSetFlagsAreValid(CORINFO_Ins
             resultflags.RemoveInstructionSet(InstructionSet_POPCNT);
 #endif // TARGET_AMD64
 #ifdef TARGET_X86
+        if (resultflags.HasInstructionSet(InstructionSet_SSE) && !resultflags.HasInstructionSet(InstructionSet_X86Base))
```
Were these hand added or did you update the templates and rerun the generator?

saucecontrol replied:
I ran the generator.  Sorry, I should have separated the generated files into their own commit to make review easier.

tannergooding commented on src/coreclr/src/jit/codegen.h:
```diff
@@ -1028,6 +1028,7 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
     void genHWIntrinsic_R_R_R_RM(
         instruction ins, emitAttr attr, regNumber targetReg, regNumber op1Reg, regNumber op2Reg, GenTree* op3);
     void genBaseIntrinsic(GenTreeHWIntrinsic* node);
+    void genX86BaseIntrinsic(GenTreeHWIntrinsic* node);
```
We could probably just merge this with `genBaseIntrinsic` (I don't have a preference on which name is taken).

We've tried to keep these compressed as more things have become table driven.

saucecontrol replied:
The main reason I added a new method was that everything in X86Base should be scalar, whereas most (all?) things in Vector are not.  I can recombine if you think we'd be able to make the codegen side table-driven, though.

tannergooding replied:
We can deal with it in a follow up PR. I'm not particularly concerned right now and it is a bigger refactoring to allow scalar intrinsics to be table driven in codegen.

tannergooding commented on src/coreclr/src/jit/emitxarch.cpp:
```diff
@@ -14818,6 +14819,8 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
             break;
 
+        case INS_bsf:
+        case INS_bsr:
```
It would be nice if we had a way to model Intel vs AMD here.

| Instruction | Intel | AMD  |
| ----------- | ----- | ---- |
| BSF         | 1.00  | 4.00 |
| BSR         | 1.00  | 5.00 |
| POPCNT      | 1.00  | 0.50 |
| LZCNT       | 1.00  | 0.50 |
| TZCNT       | 1.00  | 0.50 |

* For the non-memory encoding, AMD is actually a bit faster still
* Numbers taken from https://uops.info/table.html

tannergooding commented on src/coreclr/src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1249,6 +1253,40 @@ void CodeGen::genBaseIntrinsic(GenTreeHWIntrinsic* node)
+// Arguments:
+//    node - The hardware intrinsic node
+//
+void CodeGen::genX86BaseIntrinsic(GenTreeHWIntrinsic* node)
```
Not needed for this PR, but we allowed `HW_Category_Scalar` intrinsics to be table driven in importation.
It would be nice to do the same here in codegen...

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -67,12 +67,20 @@ public static int LeadingZeroCount(uint value)
+                // LZCNT returns index starting from MSB, whereas BSR gives the index from LSB.
+                // The XOR here is equivalent to subtraction since the BSR result is always between 0 and 31.
+                // This saves one instruction, as subtraction from constant requires either MOV/SUB or NEG/ADD.
+                return 31 ^ (int)X86Base.BitScanReverse(value);
```
Should we track this as a more general JIT optimization?

saucecontrol replied:
It only applies in very limited situations.  LHS has to be a power of 2 minus one and RHS has to be between 0 and LHS.  We know those are both true here, but I don't think the JIT could know that on its own anywhere.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/X86Base.PlatformNotSupported.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <summary>
+    /// This class provides access to the x86 base hardware instructions via intrinsics
+    /// </summary>
+    internal static class X86Base
```
@jkotas, would you say this is one that we should take to API review or leave be?

C/C++ expose intrinsics for it and it also has a 16-bit overload. Alternatively, we could propose/review general overloads that take `byte`, `sbyte`, `short`, and `ushort`.

If we aren't going to take these, we should probably leave a remark indicating as such and the reason why (they are covered by the more general use `BitOperations` methods).

jkotas replied:
I do not think these should be public APIs.

jkotas replied:
Same reason as what's discussed in https://github.com/dotnet/runtime/issues/953

tannergooding replied:
Thanks!

@saucecontrol, could you update the remarks section to indicate these are not public as they are sufficiently covered by corresponding methods in `BitOperations`?

saucecontrol replied:
Will do.  Also wasn't sure about whether to follow the abstract class pattern of the others since the class will become public if https://github.com/dotnet/runtime/issues/785 comes to be.  I assume this will have to remain standalone since we can't change the base class of the other x86 ISAs so static class is correct?

tannergooding replied:
I'm fairly certain changing the base class is fine, all members are static (and therefore also non-virtual) and so it only impacts what members are visible through IntelliSense.

saucecontrol replied:
Ah yes... I'll leave it for now.  The decision on changing the base class of the others can be taken if/when `X86Base` is made public

saucecontrol commented:
> Looks like something isn't quite hooked up right as `X86Base.IsSupported` is returning false for various tests.

I saw that `ArmBase` is explicitly added with the , even though it looks like it should already be there from the VM side:

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/compiler.cpp#L2313-L2320

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/codeman.cpp#L1485-L1487

I couldn't see any cases where X86Base wasn't working, but maybe it needs the same treatment if you're seeing a failure.

tannergooding commented:
> I couldn't see any cases where X86Base wasn't working, but maybe it needs the same treatment if you're seeing a failure.

Seeing failures in CI for x86.

saucecontrol commented:
Ah, I found one:

```
System.PlatformNotSupportedException : Operation is not supported on this platform.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/X86Base.cs(34,0): at System.Runtime.Intrinsics.X86.X86Base.X64.BitScanReverse(UInt64 value)
```

So it looks like `X86Base.X64.IsSupported` is returning true when it shouldn't.  I'll investigate

tannergooding commented on src/libraries/System.Utf8String.Experimental/src/System/Runtime/Intrinsics/Intrinsics.Shims.cs:
```diff
@@ -26,6 +26,18 @@ internal static class Vector128
+        public abstract class X64
+        {
+            public const bool IsSupported = false;
+            public static ulong BitScanForward(ulong value) => throw new PlatformNotSupportedException();
```
This should probably be `internal` to match the `X86Base`...

saucecontrol replied:
Yeah, the X86Base in this shim will always be internal so it doesn't really matter, but it wouldn't hurt to match the actual API better. Will tidy this up.

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -160,15 +180,19 @@ public static int Log2(ulong value)
                 return 63 - ArmBase.Arm64.LeadingZeroCount(value);
             }
 
+            if (X86Base.IsSupported)
+            {
+                return (int)X86Base.X64.BitScanReverse(value);
```
x86 test failures come from the incorrect `IsSupported` check here.  JIT side appears to be fine

tannergooding commented:
We need the instruction sets added here: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/compiler.cpp#L2205, in the same way as we do for ARM here: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/compiler.cpp#L2316

It looks like all the other JIT bits exist (comparing to `ArmBase` and `Sse.X64`)

saucecontrol commented:
The way I read it, the explicit addition of ArmBase is redundant because it's passed in from the VM side in the same way the higher-level ISAs are.  Since I've added X86Base unconditionally here:

https://github.com/dotnet/runtime/blob/47bc99ba0b45f8cf2144e590013c61ad911865db/src/coreclr/src/vm/codeman.cpp#L1302

It's always available unless removed based on the config knobs, just like SSE+ are.

Am I missing something there?

tannergooding commented:
> The way I read it, the explicit addition of ArmBase is redundant because it's passed in from the VM side in the same way the higher-level ISAs are. Since I've added X86Base unconditionally here:

@davidwrighton, did the base checks get inverted? That is should https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/compiler.cpp#L2205 be:
```
if (!JitConfig.EnableHWIntrinsic())
{
    // Dummy ISAs for simplifying the JIT code
    instructionSetFlags.RemoveInstructionSet(InstructionSet_Vector128);
    instructionSetFlags.RemoveInstructionSet(InstructionSet_Vector256);
}
```

You are meant to be able to disable the "baseline" as well, so you can test your software fallback.

saucecontrol commented:
The Vector ISAs don't actually exist, so they're not passed into the JIT.  Unless `JitConfig.EnableHWIntrinsic` is true, they never get added so you get the software fallbacks.

The real ISAs work the opposite way.  The VM passes them in if the hardware supports them and the JIT disables them if the config says to.

`X86Base` is working as expected with that model.  It reports `IsSupported` if COMPlus_EnableHWIntrinsic is on and not if it's not.

saucecontrol committed:
fix bad IsSupported check, PR feedback

saucecontrol commented:
It's probably a good idea to make `ArmBase` and `X86Base` handled consistently in compiler.cpp.  I actually found the explicit addition of `ArmBase` with the dummy ISAs confusing, so I'd be in favor of removing that if it's not actually necessary.

If I've misunderstood any of that, I guess we should add `X86Base` explicitly to make them match 🤷‍♂

davidwrighton commented:
@tannergooding @saucecontrol Vector64/128/256 are set by the JIT itself as Tanner noted. 

One of the changes I did not realize I made was to make ArmBase set by the VM side. One of the issues is that the VM now attempts to pass a rationalized set of instruction set values to the JIT, and not enabling ArmBase, then requires the VM side to disable all of the other instruction sets. We could tweak that if we needed to but I need to know its necessary before I futz with it.

I would avoid marking the X86Base as being implied by support for SSE, as that means if it is disabled, then SSE will be disabled.

davidwrighton commented:
And yes, the current model doesn't make sense where we explicitly set the ArmBase instruction set in both the jit and the VM side of the fence.

saucecontrol commented:
> I would avoid marking the X86Base as being implied by support for SSE, as that means if it is disabled, then SSE will be disabled.

There's presently no way to disable X86Base other than to disable HWIntrisics entirely.  Isn't that the same way ArmBase works?  Or would this be a case where disabling HWIntrinsics also disables SIMD?

> And yes, the current model doesn't make sense where we explicitly set the ArmBase instruction set in both the jit and the VM side of the fence.

Would removing it from the JIT side be ok for now, given the larger changes necessary to separate it on the VM side?

tannergooding commented:
> There's presently no way to disable X86Base other than to disable HWIntrinsics entirely. Isn't that the same way ArmBase works? Or would this be a case where disabling HWIntrinsics also disables SIMD?

That is the expectation.

The helper types and the "base" ISAs don't have their own `COMPlus` knob, instead they are controlled by the `COMPlus_EnableHWIntrinsics` knob.
They still participate in the hierarchy and so doing `EnableHWIntrinsics=0` will also disable `Sse` and others (we can't very well use `Sse` if the user has disabled `HWIntrinsics` in general, after all).

tannergooding commented:
> I would avoid marking the X86Base as being implied by support for SSE, as that means if it is disabled, then SSE will be disabled.

That is exactly what we want. Disabling `X86Base` requires disabling HWIntrinsics in general, which should also disable `SSE` (only intrinsics, not general JIT support outside HWIntrinsics of course).

davidwrighton commented:
Ah, ok then. Well, then the model should be ArmBase, and X86Base are set unconditionally by the VM side of the world. And if HWIntrinsics aren't enabled, then those two should be explicitly disabled by the JIT. @saucecontrol, would you like to do that as part of this PR?

saucecontrol commented:
> would you like to do that as part of this PR?

Sure, I don't mind cleaning that up here if I can get the model straight.

In my first attempt at this code, I actually did explicitly disable X86Base when `JitConfig.EnableHWIntrinsic` was false, but that caused some asserts in the JIT SIMD code when running with `COMPlus_EnableHWIntrinsic=0`.  IIRC, the assert was being hit on `Vector<T>.Create()`.  It seems like this might have been a side effect of SSE+ being implicitly disabled with X86Base.  I found that by leaving X86Base enabled, everything just magically worked, which I assume is because the `IsSupported` checks returned false with HWIntrinsics off but the higher ISAs were still available in the SIMD context.

The interaction between the HWIntrinsic and SIMD logic in the JIT and their dependence on the ISA checks is still a bit of a mystery to me, so I'm not sure how to resolve that... unless it's simply a matter of not making SSE depend on X86Base.  I don't know if the same would apply on the ARM side since I don't know how much of the SIMD code is wired up for ARM.

saucecontrol commented:
Ok, I was able to remove X86Base when `!JitConfig.EnableHWIntrinsic` as long as I also remove the implication X86Base->SSE from InstructionSetDesc.txt.  Leaving the implication there and removing the base ISA brought back the assert on `Vector<T>.Create()`.

So that leaves the question of what happens with ARM if I remove ArmBase.  Since ArmBase implies AdvSimd, and I believe the `Vector<T>` code relies on AdvSimd, I think that may result in the same assert I'm seeing on the xArch side. @tannergooding does that sound right?

If that's right, do we want to break the implications, or do we want to have the base ISAs unconditionally enabled on the JIT side, counting on the fact that EnableHWIntrisic=false will cause their `IsSupported` properties to return false?

saucecontrol commented:
Actually, I think I see the problem...

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/ee_il_dll.cpp#L345-L357

Here, the `Vector<T>` size is entirely based on the individual JIT ISA config knobs. With COMPlus_EnableHWIntrinsic=0, we have no X86Base and hence no AVX2, but none of the ISA knobs are disabled, so it still gets the AVX2 vector size.  But then AVX2 is disabled in `compSupportsISA`.

That whole thing is sketchy anyway since you can still get SSE2 code with COMPlus_EnableSSE=0 because the SIMD code selectively ignores the ISA knobs.

saucecontrol committed:
remove redundant ArmBase additions, add comment

tannergooding commented:
The issue here looks to come from the weird coupling of `FEATURE_SIMD` and `FEATURE_HWINTRINSIC` we had to do for back-compat.

`FEATURE_SIMD` has existed since ~2014 and had already exposed the `COMPlus_EnableAVX=0` and `COMPlus_EnableSSE3_4` flags which impacted the use of the VEX encoding and the size/codegen of `Vector<T>`.
We introduced `FEATURE_HWINTRINSIC` in .NET Core 3 and needed a way to maintain the existing flag behavior while also allowing users to disable the other ISAs, so we reused the existing flag and introduced similarly named `COMPlus_EnableISA` flags that were hierarchical in nature.

However, we (at the time) didn't want users disabling `HWIntrinsics` to also disable `Vector<T>` support since they are distinct codepaths: <https://github.com/dotnet/coreclr/pull/24649>

> That whole thing is sketchy anyway since you can still get SSE2 code with COMPlus_EnableSSE=0 because the SIMD code selectively ignores the ISA knobs.

They are respected for everything except SSE/SSE2 code (which is the runtime baseline and necessary for some non SIMD code) and so the knobs aren't ignored entirely. That is, `COMPlus_EnableAVX=0` will still disable the VEX encoding and force `Vector<T>` to be 16-bytes.

With issues like https://github.com/dotnet/runtime/issues/956 and investigations around things like doing the replacement in importation (https://github.com/dotnet/runtime/issues/9766#issuecomment-605300787) the features will become more coupled and we will likely have to have `EnableHWIntrinsic` impact the size of `Vector<T>`, so maybe it is okay to just do this now.
* We probably shouldn't make that change as part of this PR, if it is being made at all. I'll open an issue for further discussion and tag the relevant people.
  I don't think ARM will hit the same issue since it doesn't support variable sized `Vector<T>` right now

saucecontrol commented:
So there are two options here if we want to get the behavior (and code) consistent between xArch and ARM:

1) Let the base ISAs flow in from the VM and preserve the values on the JIT side.  The EnableHWIntrinsic knob still allows disabling them because their intrinsic mappings are disabled (i.e. `IsSupported` is false).
2) Explicitly disable the base ISAs on the JIT side when EnableHWIntrinsic is off.  This prevents us from making SSE depend on X86Base because of the SIMD Vector size issue, but ARM should work as is.

My personal preference is option 1, so I went ahead and pushed a commit that does that.  If for some reason we ever wanted to be able to disable the base ISAs independent of the HWIntrinsic knob, we could address the dependency issue at that time.  As you say, this will need to be revistited if the SIMD intrinsics become dependent on HWIntrinsics anyway.

saucecontrol commented:
I haven't been able to figure out the remaining x86 test failures.  With BMI1 and LZCNT enabled on a compatible processor, this should be a zero-diff change on the JIT side, and only the subtraction->xor change to `BitOperations.Log2` on the managed side.  However, I have a `System.Memory` test that fails reliably on x86 with my PR build.  That particular test only makes use of `BitOperations.TrailingZeroCount`, and disabling BMI1 does switch the TZCNT to BSF as expected... then it fails in exactly the same way.

I tried running a jit-diff against the base of my branch, but I get an error about the JIT being the wrong version (is that because of the GUID change?).  Running the test with COMPlus_JitDisasm, the jitted code is identical right up to the point the baseline completes the test and the PR build throws an exception.

I'll do another debug session on it tomorrow, but for now I'm stumped.

saucecontrol commented:
Ah, looks like it must be a crossgen issue.  The IL-only corelib passes the tests.

tannergooding commented:
CC. @davidwrighton. Possibly a side-effect of the recent changes?

saucecontrol commented:
Certainly appears to be a bad interaction between my change and the ISA support change in crossgen…

My changes to that side are almost all straight out of the ThunkGenerator util, but I could have missed something.

Couple of observations that might help narrow it down:

1) The x86 tests that fail are different between checked and release libraries builds.  I noticed that behavior in the CI results but then it repros locally.  Debug libraries builds pass, and I can make the others pass by dropping the pre-R2R corelib into my testhost.
2) The R2R corelib works fine in x64 under any config.  Considering the X86Base_X64 ISA is dependent on X86Base, I would have expected an x86 failure to flow through to x64.

davidwrighton commented:
1. I don't see anything wrong with the integration with my changes.
2. You did however miss handling the logic in `FilterNamedIntrinsicMethodAttribs` in zapinfo.cpp. That will have some effect on the crossgen of System.Private.CoreLib, but I wouldn't expect it to actually break things, just to make applications run inexplicably slower than necessary.

saucecontrol commented:
@davidwrighton thanks for the second look.  I did totally miss the ISA section in zapinfo.cpp.  I'll get that fixed up and then dig into the generated code in the x86 R2R corelib and see what I can see.

One more question, though... in this section:

https://github.com/dotnet/runtime/blob/ef54ab092a6876c2efc6a34d8f0203d98d1868f5/src/coreclr/src/zap/zapinfo.cpp#L2170-L2172

shouldn't those ISA compares in the outer `if` be comparing on `isaName` rather than `className`?  There's no way for the `X64` comparison below to get caught as it is.  I wonder if that might allow an x64 instruction to get compiled into an x86 image?

saucecontrol committed:
add X86Base to zapinfo

saucecontrol commented:
Can't fully explain why, but that seems to have fixed it up.  I get clean test runs locally with the x86 R2R corelib.

davidwrighton commented:
I've filed #34826 based on your comment. It looks like a bug to me.

Unfortunately, if that fixed things it means that something is wrong with having the IsSupported check be an actual call instead of a jit intrinsic I think. That's very odd, and I want you to track why that is happening before I think its reasonable to sign off on the change.

tannergooding commented:
> I've filed #34826 based on your comment. It looks like a bug to me.

It looks to be a bug to me as well, and I linked to the original PR where I made the changes

saucecontrol commented:
Well, looks there's still a problem: https://helix.dot.net/api/2019-06-17/jobs/81c16505-2592-44b3-a03d-b8b70965016d/workitems/System.Runtime.Extensions.Tests/console

The results there appear to be garbage but are all convieniently in the range of valid BSF results. The only thing I can think of that would cause that is the emitter encoding the wrong operand.  @tannergooding any ideas how that could happen?

tannergooding commented:
Will try to repro locally. It is likely only showing on Mac because there are some machines in the pool which don't support BMI1 and so they are falling back to BSR/BSF.
The other machines all support BMI1 and so they never hit this code path.

saucecontrol commented:
BMI1 is disabled in crossgen, so this should actually be switching from the software fallback to BSF on all xArch.  I've verified that BSF is encoded correctly in the majority of cases, but I suspect that in some cases the operand register is wrong.

My theory for why the changes to recognize the `IsSupported` check in crossgen fixed the previous failures on x86 is just that the code was simplified enough that the register flipped back to the correct one.  I haven't been able to find an exact case where it was broken, but I saw an example where the codegen went from using a combination of eax,edx,edi to just eax.  Harder to get it wrong if there's only one being used...

saucecontrol commented:
One more interesting thing I noticed in examining the crossgen asm…

Since BMI1 is disabled in crossgen, `BitOperations.TrailingZeroCount` goes from unconditionally using the software fallback to unconditionally using BSF with this change.  However, because LZCNT is enabled in crossgen but can't be guaranteed to be present on the hardware, both the `IsSupported` and the `LeadingZeroCount` are emitted as regular method calls with a fallback to BSR if `Lzcnt.IsSupported` returns false.  It might be better to disable LZCNT in crossgen so it just emits BSR unconditionally.  Even if BSR is slower on some processors, it's better than keeping the two method calls to use LZCNT.

That's assuming one of us figures out how to get BSF/BSR working, of course 😆

tannergooding commented:
I can repro on Windows, with BMI1, ReadyToRun, and TieredCompilation disabled.

The JIT thinks it is emitting `bsf ecx, dword ptr [rbp+10H]`, but the bytes (`0FBD4D10`) are actually: `bsr ecx, dword ptr [rbp + 0x10]`

tannergooding commented:
Looks like the emitter is changing it here: <https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/emitxarch.cpp#L10216>

I believe it is attempting to set the 'w' bit (`B.1.4.3 Encoding of Operand Size (w) Bit`)

tannergooding commented:
`B.2 GENERAL-PURPOSE INSTRUCTION FORMATS AND ENCODINGS FOR NON-64-BIT MODES` and `B.2.1 General Purpose Instruction Formats and Encodings for 64-Bit Mode` give a list of instructions to which the `w` bit and the `s` bit actually exist. This is something that could probably be encoded in the `INS_FLAGS` section: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/instrsxarch.h, but that is a more complicated change.

@CarolEidt, @briansull; What would be the preferred way of resolving the issue for this PR?
I'd imagine we just want something smaller here and potentially make a more involved fixup in a separate PR...

saucecontrol committed:
update InstructionSetDesc.txt, run ThunkGenerator

saucecontrol committed:
implement X86Base, add BSF and BSR intrinsics

saucecontrol committed:
add BSF and BSR fallbacks in BitOperations

saucecontrol committed:
disable 'w' size bit for BSF and BSR

saucecontrol commented:
Rebased on master to resolve the conflicts in zapinfo.cpp and split the first commit to separate out the generated files from manually updated.

I special-cased the 'w' bit handling for `BSF` and `BSR` for now and left a TODO comment for later cleanup.

The current handling of the 'w' bit is inconsistent in general, and I may have have spotted a bug in the `CMOVcc` encoding.  In one case, CMOV instructions are correctly exempted from the 'w' bit encoding:
https://github.com/dotnet/runtime/blob/19095805c8c5e1d05fb45b12c5908787a13572fb/src/coreclr/src/jit/emitxarch.cpp#L10216-L10220

And in another they're explicitly included:
https://github.com/dotnet/runtime/blob/19095805c8c5e1d05fb45b12c5908787a13572fb/src/coreclr/src/jit/emitxarch.cpp#L10680-L10686

saucecontrol committed:
formatting

tannergooding reviewed (approved):
LGTM.

CC. @dotnet/jit-contrib for additional sign-off

CarolEidt reviewed (approved):
LGTM, though I'd like @davidwrighton to have a look for the ISA support.

tannergooding commented:
I logged https://github.com/dotnet/runtime/issues/35305 to track the emitter's handling of the 'w' and 's' encoding bits.

saucecontrol committed:
update to new hwintrinsiclist table format

saucecontrol commented:
Rebased on master to bring in https://github.com/dotnet/runtime/pull/35364, and updated to the new intrinsiclist table layout

saucecontrol committed:
formatting

saucecontrol committed:
formatting

saucecontrol commented:
Test failures related to `Array.Sort` gave me a scare for a second, but those were apparently https://github.com/dotnet/runtime/issues/35438

CarolEidt commented:
ping @davidwrighton - could you have a look at the ISA handling here?

davidwrighton reviewed (approved):
This looks good. I believe the new instruction set should be enabled by the existing pathways in crossgen2, so no additional crossgen2 change should be needed.

saucecontrol committed:
re-run thunkgenerator

saucecontrol commented:
Rebased again to get https://github.com/dotnet/runtime/pull/35504, and re-ran thunkgenerator

saucecontrol committed:
fix GUID conflict from https://github.com/dotnet/runtime/pull/35552

tannergooding commented:
Thanks for the contribution @saucecontrol!
